packages/flutter/lib/src/widgets/virtual_viewport.dart

```diff
@@ -162,3 +148,84 @@ abstract class VirtualViewportElement<T extends VirtualViewport> extends RenderO
     renderObject.remove(child);
   }
 }
+
+abstract class VirtualViewportIterableMixin extends VirtualViewport{
+  Iterable<Widget> get children;
+
+  _IterableWidgetProvider _createWidgetProvider() => new _IterableWidgetProvider();
+}
+
+class _IterableWidgetProvider extends _WidgetProvider {
+  int _length;
+  Iterator<Widget> _iterator;
+  List<Widget> _widgets;
+
+  void didUpdateWidget(VirtualViewportIterableMixin oldWidget, VirtualViewportIterableMixin newWidget) {
+    if (oldWidget == null || newWidget.children != oldWidget.children) {
+      _iterator = null;
+      _widgets = <Widget>[];
+      _length = newWidget.children.length;
+    }
+  }
+
+  int get virtualChildCount => _length;
+
+  void prepareChildren(VirtualViewportElement context, int base, int count) {
+    int limit = base < 0 ? _length : math.min(_length, base + count);
+    if (limit <= _widgets.length)
+      return;
+    VirtualViewportIterableMixin widget = context.widget;
+    if (widget.children is List<Widget>) {
+      _widgets = widget.children;
+      return;
+    }
+    _iterator ??= widget.children.iterator;
+    while (_widgets.length < limit) {
+      bool moved = _iterator.moveNext();
+      assert(moved);
+      Widget current = _iterator.current;
+      assert(current != null);
+      _widgets.add(current);
+    }
+  }
+
+  Widget getChild(int i) => _widgets[(i % _length).abs()];
+}
+
+typedef List<Widget> ItemListBuilder(BuildContext context, int start, int count);
+
+abstract class VirtualViewportLazyMixin extends VirtualViewport {
```

Comment:
Yeah, I fought with the type system on how to express this.  I suspect it will need another iteration to get the itemBuilder cache invalidation perfect.  The problem is that it's a two dimensional space (source of child widgets and layout model), which doesn't match up with object-oriented programming very well.